pkg/receive/handler.go

```diff
@@ -482,6 +548,62 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 	h.writeSamplesTotal.WithLabelValues(strconv.Itoa(responseStatusCode), tenant).Observe(float64(totalSamples))
 }
 
+// QueryMetaMonitoring queries any Prometheus Query API compatible meta-monitoring
+// solution with the configured query for getting current HEAD series of all tenants.
+// It then populates tenantCurrentSeries map with result.
+func (h *Handler) QueryMetaMonitoring(ctx context.Context) error {
+	c := promclient.NewWithTracingClient(h.logger, h.metaMonitoringClient, httpconfig.ThanosUserAgent)
+
+	vectorRes, _, err := c.QueryInstant(ctx, h.options.MetaMonitoringUrl, h.options.MetaMonitoringLimitQuery, time.Now(), promclient.QueryOptions{})
+	if err != nil {
+		h.metaMonitoringErr.Inc()
+		return err
+	}
+
+	level.Debug(h.logger).Log("msg", "successfully queried meta-monitoring", "vectors", len(vectorRes))
+
+	// Construct map of tenant name and current HEAD series.
+	for _, e := range vectorRes {
+		for k, v := range e.Metric {
+			if k == "tenant" {
+				h.tenantCurrentSeriesMap[string(v)] = float64(e.Value)
+				level.Debug(h.logger).Log("msg", "tenant value queried", "tenant", string(v), "value", e.Value)
+			}
+		}
+	}
+
+	return nil
+}
+
+// isUnderLimit ensures that the current number of active series for a tenant does not exceed given limit.
+// It does so in a best-effort way, i.e, in case meta-monitoring is unreachable, it does not impose limits.
+// TODO(saswatamcode): Add capability to configure diff limits for diff tenants.
+func (h *Handler) isUnderLimit(tenant string, logger log.Logger) (bool, error) {
+	if h.options.MaxPerTenantLimit == 0 || h.options.MetaMonitoringUrl.Host == "" {
+		return true, nil
+	}
+
+	// In such limiting flow, we ingest the first remote write request
+	// and then check meta-monitoring metric to ascertain current active
+	// series. As such metric is updated in intervals, it is possible
+	// that Receive ingests more series than the limit, before detecting that
+	// a tenant has exceeded the set limits.
+	v, ok := h.tenantCurrentSeriesMap[tenant]
+	if !ok {
```

Comment:
In this case, are we taking into consideration this could be the first request from a new tenant? This should not be considered an error I think.

Reply:
So this error is not explicit (due to best-effort limiting), we only log this error. There doesn't seem to be a neat way of detecting that this is the first request of a tenant via Handler. Also, in case a tenant sends multiple request before we have had a chance to query meta-monitoring, all those requests would log errors, even if we detect first one.

WDYT would be a good way of handling this? 🙂

Reply:
Hm that is true that we could receive multiple requests before the 15 second interval from a new tenant, feels like that's simply shortcoming of the meta-monitoring approach. Either way, are there other cases where a tenant would be missing from the map, apart from the case of new tenant? In that case I'd not consider it an error but simply return `true` without error and let the tenant have it "for free" before we query meta-monitoring.

 One thing that comes to my mind would be using a channel to signal to the querying goroutine to fire off, once we detect a request from new tenant has been ingested? Might be easier said than done.

Reply:
Will try this out. But another reason why a tenant may not be in the map is that meta-monitoring does not have metrics for one instance of a Receive, thereby completely missing a few tenants. In which case such an error log could be a clue. WDYT?

Reply:
Hm right I have not thought about hard tenancy :thinking:. Could we detect within meta-monitoring if we have metrics for expected number of receiver instances and if not warn about it?